pkg/tfbridge/validate_input_types_test.go

```diff
@@ -0,0 +1,1883 @@
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			name:     "object_multi_type_enum_no_failure",
+			typeRef:  "ObjectMultiType",
+			typeName: "object",
+			input: resource.NewObjectProperty(resource.NewPropertyMapFromMap(map[string]interface{}{
+				"prop": "foo",
+			})),
+			types: map[string]pschema.ComplexTypeSpec{
+				"pkg:index/type:EnumType": {
+					ObjectTypeSpec: pschema.ObjectTypeSpec{
+						Type: "string",
+					},
+					Enum: []pschema.EnumValueSpec{
+						{Name: "Value1", Value: "v1"},
+						{Name: "Value2", Value: "v2"},
+					},
+				},
+				"pkg:index/type:ObjectMultiType": {
+					ObjectTypeSpec: pschema.ObjectTypeSpec{
+						Type: "object",
+						Properties: map[string]pschema.PropertySpec{
+							"prop": {
+								TypeSpec: pschema.TypeSpec{
+									OneOf: []pschema.TypeSpec{
+										{
+											Type: "object",
+											Ref:  "#/types/pkg:index/type:EnumType",
+										},
+										{
+											Type: "number",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			name:     "object_multi_type_success_number",
+			typeRef:  "ObjectMultiType",
+			typeName: "object",
+			input: resource.NewObjectProperty(resource.NewPropertyMapFromMap(map[string]interface{}{
+				"prop": 1,
+			})),
+			types: map[string]pschema.ComplexTypeSpec{
+				"pkg:index/type:ObjectMultiType": {
+					ObjectTypeSpec: pschema.ObjectTypeSpec{
+						Type: "object",
+						Properties: map[string]pschema.PropertySpec{
+							"prop": {
+								TypeSpec: pschema.TypeSpec{
+									OneOf: []pschema.TypeSpec{
+										{
+											Type: "string",
+										},
+										{
+											Type: "number",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			name:     "object_multi_type_success_object",
+			typeRef:  "ObjectMultiType",
+			typeName: "object",
+			input: resource.NewObjectProperty(resource.NewPropertyMapFromMap(map[string]interface{}{
+				"prop": map[string]string{"foo": "bar"},
+			})),
+			types: map[string]pschema.ComplexTypeSpec{
+				"pkg:index/type:ObjectMultiType": {
+					ObjectTypeSpec: pschema.ObjectTypeSpec{
+						Type: "object",
+						Properties: map[string]pschema.PropertySpec{
+							"prop": {
+								TypeSpec: pschema.TypeSpec{
+									OneOf: []pschema.TypeSpec{
+										{
+											Type: "string",
+										},
+										{
+											Type: "object",
+											AdditionalProperties: &pschema.TypeSpec{
+												Type: "string",
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			name:     "oneof_object_multi_type_failure_object",
+			typeRef:  "ObjectMultiType",
+			typeName: "object",
+			input: resource.NewObjectProperty(resource.NewPropertyMapFromMap(map[string]interface{}{
+				"prop": []interface{}{"foo"},
+			})),
+			failures: []TypeFailure{
+				{
+					Reason:       "expected string type, got [] type",
```

Comment:
If we can't resolve the inner type here, can we call this a list type?
```suggestion
					Reason:       "expected string type, got list type",
```

Reply:
To generate this I'm just using `PropertyValue.TypeString()` and it unfortunately uses `[]` as the value for array.

https://github.com/pulumi/pulumi/blob/ff8b280cc962b66dcc0c7756575b4469a6c0cf7d/sdk/go/common/resource/properties.go#L558-L560

I can definitely customize this to return `list` instead if that is preferred.

Reply:
If it's coming from upstream, we can leave it for now.